src/main/java/com/n1analytics/paillier/EncodedNumber.java

```diff
@@ -129,87 +127,67 @@ public EncodedNumber checkSameContext(EncodedNumber other)
-    return decode().decodeApproximateBigInteger();
-  }
-
-  /**
-   * Decodes to a {@code BigInteger} representation.
+   * Decodes to a {@code BigInteger} representation. See
+   * {@link com.n1analytics.paillier.PaillierContext#decodeBigInteger(EncodedNumber)} for details.
    *
    * @return the decoded number.
    * @throws ArithmeticException if this {@code EncodedNumber} cannot be represented as a {@code BigInteger}.
    */
   public BigInteger decodeBigInteger() throws ArithmeticException {
-    return decode().decodeBigInteger();
-  }
-
-  /**
-   * Decodes to the approximated {@code double} representation.
-   * @return the decoded number.
-   */
-  public double decodeApproximateDouble() {
-    return decode().decodeApproximateDouble();
+    return context.decodeBigInteger(this);
   }
 
   /**
-   * Decodes this {@code EncodedNumber} to a {@code double} representation. Throws an ArithmeticException
-   * if this {@code EncodedNumber} cannot be represented as a valid {@code double}.
+   * Decodes this {@code EncodedNumber} to a {@code double} representation. See
+   * {@link com.n1analytics.paillier.PaillierContext#decodeDouble(EncodedNumber)} for details.
    *
    * @return the decoded number.
    * @throws ArithmeticException if this {@code EncodedNumber} cannot be represented as a valid {@code double}.
    */
   public double decodeDouble() throws ArithmeticException {
-    return decode().decodeDouble();
+    return context.decodeDouble(this);
   }
 
   /**
-   * Decodes this {@code EncodedNumber} to an approximated {@code long} representation. If the number
-   * cannot be represented exactly as a {@code long}, it is converted to the {@code long} representation
-   * of the lowest 64 bits.
-   *
-   * @return the decoded number.
-   */
-  public long decodeApproximateLong() {
-    return decode().decodeApproximateLong();
-  }
-
-  /**
-   * Decodes this {@code EncodedNumber} to a {@code long} representation. Throws an ArithmeticException
-   * if this cannot be represented as a valid {@code long}.
+   * Decodes this {@code EncodedNumber} to a {@code long} representation. See
+   * {@link com.n1analytics.paillier.PaillierContext#decodeLong(EncodedNumber)} for details.
    *
    * @return the decoded number.
    * @throws ArithmeticException if this cannot be represented as a valid {@code long}.
    */
   public long decodeLong() throws ArithmeticException {
-    return decode().decodeLong();
+    return context.decodeLong(this);
   }
 
   /**
    * Re-encodes this number with the specified context.
    *
-   * @param context the context to re-encode with.
+   * @param context the context to re-encode this number with.
    * @return the re-encoded number.
    */
   public EncodedNumber changeContext(PaillierContext context) {
-    return context.encode(decode());
+    if(exponent < 0)
+      return context.encode(this.context.decodeDouble(this));
+    else
+      return context.encode(this.context.decodeBigInteger(this));
   }
```

Comment:
- I find it unintuitive why there are decodeDouble and decodeBigInteger depending on the exponent. A short explanation would help.
- You potentially loose **a lot** of information, as you squeeze your n-bit significand into a 53 bit double significand. We should at least add a warning here.
- even better: instead of decoding to a double, we should decode to a BigDecimal, this way we minimise information loss
- I think our coding convention ask for curly brackets everywhere.